compiler/rustc_hir_analysis/src/check/compare_impl_item.rs

```diff
@@ -831,7 +831,9 @@ impl<'tcx> TypeFolder<TyCtxt<'tcx>> for ImplTraitInTraitCollector<'_, 'tcx> {
 
     fn fold_ty(&mut self, ty: Ty<'tcx>) -> Ty<'tcx> {
         if let ty::Alias(ty::Projection, proj) = ty.kind()
-            && self.interner().def_kind(proj.def_id) == DefKind::ImplTraitPlaceholder
+            && (self.interner().def_kind(proj.def_id) == DefKind::ImplTraitPlaceholder
+                || self.interner().lower_impl_trait_in_trait_to_assoc_ty()
+                    && self.interner().def_kind(proj.def_id) == DefKind::AssocTy)
```

Comment:
Might want to move this into a helper function on tcx. We'll be doing this all over the place probably.

Reply:
Also, this will match on _every_ associated type in a type signature, I think? I don't think that's what we want.

Reply:
checking if something is an old-style RPITIT def id (by checking def kind is ImplTraitPlaceholder), or a new style RPITIT def id (by checking opt_rpitit_info)

Reply:
I've added a function for new style but maybe you want a function that integrate both.

Reply:
We want something that will check both because we would ideally be replacing existing checks for DefKind::ImplTraitPlaceholder to one single helper function that works correctly depending on if the Z flag is active or not